pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,138 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Package contains all the metadata of the package and drives package searching and browsing,
+// and support the fork/rebase upgrade workflow. It can be used by something like an app registry.
+type Package struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// The name of the package.
+	// The name of an individual instance should live in ObjectMeta.Name.
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Description of the package.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// An pointer to the icon.
+	Icon string `json:"icon,omitempty" yaml:"icon,omitempty"`
+
+	// Search keywords for the package.
+	Keywords []string `json:"keywords,omitempty" yaml:"keywords,omitempty"`
+
+	// Homepage of the application package.
+	Home string `json:"home,omitempty" yaml:"home,omitempty"`
+
+	// Source specifies the upstream URL, e.g. https://github.com/foo/bar.git, file://host/path, etc.
+	// hosting the resource files specified in Base and Overlays.
+	// This is useful in the fork/rebase workflow.
+	Source string `json:"source,omitempty" yaml:"source,omitempty"`
+
+	// Version of the package.
+	PackageVersion string `json:"packageVersion,omitempty" yaml:"packageVersion,omitempty"`
+}
+
+// Manifest has all the information to expand of generate the k8s api resources.
+// It can be used by kubectl or some other tooling.
+// A manifest could be either a Base or an Overlay.
+type Manifest struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// TODO: figure out if we need field ManifestVersion.
+	// See: https://github.com/kubernetes/kubernetes/pull/52570/files/3eea91793dfbc3fdb0799589fac3790c4cde58a4#r140391019
+	// Version of the manifest.
+	// ManifestVersion string `json:"manifestVersion,omitempty" yaml:"manifestVersion,omitempty"`
+
+	// NamePrefix will prefix the names of all resources mentioned in the manifest
+	// including generated configmaps and secrets.
```

Comment:
Wow! Will it also rewrite anywhere where (eg) services are referenced as hostnames in config and command line flags?

I guess what I'm trying to say is: why isn't this done via the overlay mechanism?  And .. doesn't it _need_ to be done via the overlay mechanism because changing names will commonly involve more than a naive objectmeta rewrite?

Reply:
Because overlays are overkill to just do this, and this is by far the most common kind of customization performed. Discussed originally in #1698.

But, yes, this mechanism is the trickiest part of my proposal, and a proposal about this specific feature will be written.

All of the other features are modest extensions of existing functionality.

Reply:
Another reason I see is we need to match the Base API resource when applying an Overlay base on <Group, Version, Kind, Name>.
Otherwise we will need some rule for the resource name in Overlay in order to find its corresponding Base.

Reply:
Great observation.

Reply:
Re. services, I want the explicit links proposal to move forward (though it may need some adjustments):
https://github.com/kubernetes/community/pull/176